Console/Command/BatchingOptimizeCommand.php

```diff
@@ -0,0 +1,366 @@
+            self::LARGE_SAMPLE_SIZE :
+            self::DEFAULT_SAMPLE_SIZE;
+        $simpleSampleSize = (int)round($sampleSize * ($this->storeCounts[$storeId]['simple_percentage'] / 100));
+        $complexSampleSize = (int)round($sampleSize * ($this->storeCounts[$storeId]['complex_percentage'] / 100));
+
+        $this->storeCounts[$storeId]['simple_sample_size'] = $simpleSampleSize;
+        $this->storeCounts[$storeId]['complex_sample_size'] = $complexSampleSize;
+
+        $this->storeCounts[$storeId]['sample'] = array_merge(
+            $this->getProductsSizes($simpleProducts, $simpleSampleSize),
+            $this->getProductsSizes($complexProducts, $complexSampleSize)
+        );
+    }
+
+    /**
+     * @param int $storeId
+     * @param array $productTypes
+     * @return Collection
+     */
+    protected function getProductsCollectionForStore(int $storeId, array $productTypes = []): Collection
+    {
+        $onlyVisible = !$this->configHelper->includeNonVisibleProductsInIndex();
+        $collection = $this->productHelper->getProductCollectionQuery($storeId, null, $onlyVisible);
+        if (count($productTypes) > 0) {
+            $collection->addAttributeToFilter('type_id', ['in' => $productTypes]);
+        }
+
+        // Randomize the results to get a more "diverse" sample
+        $collection->getSelect()->orderRand();
+
+        return $collection;
+    }
+
+    /**
+     * @param Collection $products
+     * @param int $sampleSize
+     * @return array
+     * @throws LocalizedException
+     * @throws NoSuchEntityException
+     * @throws DiagnosticsException
+     * @throws AlgoliaException
+     */
+    protected function getProductsSizes(Collection $products, int $sampleSize): array
+    {
+        $stats = [];
+        $limit = 0;
+
+        foreach ($products as $product) {
+            if ($limit >= $sampleSize) {
+                break;
+            }
+
+            $serializedRecord = json_encode($this->recordBuilder->buildRecord($product));
+
+            if (function_exists('mb_strlen')) {
+                $size = mb_strlen($serializedRecord, '8bit');
+            } else {
+                $size = strlen($serializedRecord);
+            }
+
+            $stats[$product->getSku()] = $size;
+            $limit++;
+        }
+
+        return $stats;
+    }
+
+    /**
+     * @param array $sizes
+     * @return int
+     */
+    protected function getSizeAverage(array $sizes): int
+    {
+        return (int) round(array_sum(array_values($sizes)) / count($sizes));
+    }
+
+    /**
+     * @param int $averageSize
+     * @return int
+     */
+    protected function getEstimatedMaxBatchCount(int $averageSize): int
+    {
+        return (int) round(self::MAX_BATCH_SIZE / $averageSize);
+    }
+
+    /**
+     * @param array $sizes
+     * @param int $averageSize
+     * @return float
+     */
+    protected function getStandardDeviation(array $sizes, int $averageSize): float
+    {
```

Comment:
I'm concerned about possible division by zero errors.
Can we guard these methods with something like?
```
if (count($sizes) <= 1) {
   return 0.0;
}
```